src/tsolvers/axdiffsolver/ArraySolver.cc

```diff
@@ -0,0 +1,811 @@
+/*
+ *  Copyright (c) 2021-2022, Martin Blicha <martin.blicha@gmail.com>
+ *
+ *  SPDX-License-Identifier: MIT
+ *
+ */
+
+#include "ArraySolver.h"
+
+#include "TreeOps.h"
+
+#include <numeric>
+
+static SolverDescr descr_ax_solver("Array Solver", "Solver for Theory of Arrays");
+
+ArraySolver::ArraySolver(Logic & logic, Egraph & egraph, SMTConfig & config) :
+    TSolver((SolverId) descr_ax_solver, (const char *) descr_ax_solver, config),
+    logic(logic),
+    egraph(egraph)
+    { }
+
+ArraySolver::~ArraySolver() = default;
+
+void ArraySolver::clearSolver() {
+    clear();
+    TSolver::clearSolver();
+}
+
+bool ArraySolver::assertLit(PtAsgn literal) {
+    if (logic.isEquality(literal.tr)) {
+        setPolarity(literal.tr, literal.sgn);
+        assertedLiterals.push(literal);
+        if (literal.sgn == l_True) { // Strong equivalence context has changed -> reset
+            clear();
+        } else if (literal.sgn == l_False) {
+            // For asserted disequality check current read-over-weak-eq lemmas to see if any is now completly falsified
+            for (auto & lemma : lemmas) {
+                auto & undecided = lemma.undecidedEqualities;
+                auto it = undecided.find(literal.tr);
+                if (it != undecided.end()) {
+                    undecided.erase(it);
+                }
+                if (undecided.empty()) {
+                    has_explanation = true;
+                    computeExplanation(lemma.equality);
+                    return false;
+                }
+            }
+        }
+    }
+    return true;
+}
+
+void ArraySolver::pushBacktrackPoint() {
+    backtrack_points.push(assertedLiterals.size_());
+    TSolver::pushBacktrackPoint();
+}
+
+void ArraySolver::popBacktrackPoint() {
+    clear();
+    assert(backtrack_points.size() > 0);
+    auto lastSize = backtrack_points.last();
+    backtrack_points.pop();
+    while (assertedLiterals.size_() > lastSize) {
+        auto lit = assertedLiterals.last();
+        assertedLiterals.pop();
+        clearPolarity(lit.tr);
+    }
+
+    TSolver::popBacktrackPoint();
+}
+
+void ArraySolver::popBacktrackPoints(unsigned int i) {
+    clear();
+    TSolver::popBacktrackPoints(i);
+}
```

Comment:
Because `TSolver::popBacktrackPoints` is going to go `popBacktrackPoint`.
We could actually most likely remove the call to `clear` from here. I think I put it there just as a precaution.

At some point, it might be good to revisit also the design of these backtrack points in TSolvers.